File: src/2025/07/main.cpp
#include "hash.hpp"
#include "tokenizer.hpp"
#include <libapi.hpp>
#include <ranges>
#include <unordered_set>


ull part1(vector<vector<char>>& grid, ull i, ull j, unordered_map<pair<ull, ull>,ull, StaticPairHash<ull, ull>>& dp){
  while(i+1<grid.size() and grid[i+1][j] != '^'){
    i+=1;
  }

  if(i==grid.size()-1){
    return 1;
  }
  
  if(dp.contains({i,j})){
    return dp[{i,j}];
  }

  ull left = 0;
  ull right = 0;

  if(j-1<grid[0].size()){
    left = part1(grid, i+1, j-1, dp);
  } 
  if(j+1<grid[0].size()){
    right = part1(grid, i+1, j+1, dp);
  }
  
  dp[{i,j}] = left + right;
  return  left + right;
}

int main(){
  ifstream file = read_input("2025", "07", INPUT);
  vector<vector<char>> grid;
  auto dp = unordered_map<pair<ull, ull>, ull, StaticPairHash<ull, ull>>();
  
  for(string line; getline(file, line);){
    vector<char> row;
    for(char e: line){
      row.push_back(e); 
    }
    grid.push_back(row);
  }

  ull j = 0;
  for(auto [i,v]: srv::enumerate(grid[0])){
    if(v == 'S'){
      j = i;
      break;
    }
  }


  ull res = part1(grid,0,j, dp);
  println("{}", res);
}

File: src/2025/07/main.py


def part1(grid: list[list[str]], pos: tuple[int, int])->int:

    i,j = pos
    seen = {}

    def rec(i:int,j:int)->int:

        while i+1<len(grid) and grid[i+1][j] != "^":
            i+=1

        if i == len(grid)-1:
            return 0
        else:
            if (i,j) in seen:
                return 0
            seen[(i,j)] = 0
            print("seen")
            print(seen)
            left: int = part1(grid, (i,j-1)) if j-1>-1 else 0
            right: int = part1(grid, (i,j+1)) if j+1<len(grid[0]) else 0
            return 1+left+right
        
    return rec(i,j)
    



if __name__ == "__main__":
    input = open("inputs/2025/07/input.txt", "r")
    sample = open("inputs/2025/07/sample.txt", "r")
    
    grid: list[list[str]] =[]
    file = sample

    for line in file:
        grid.append(list(line.strip()))
    
    j=0
    while j<len(grid[0]):
        if grid[0][j] == "S":
            break
        else:
            j+=1
    
    print(part1(grid, (0,j)))

